include/proxy-wasm/context.h

```diff
@@ -0,0 +1,731 @@
+   */
+  virtual void sendLocalResponse(uint64_t /* response_code */, string_view /* body */,
+                                 Pairs /* additional_headers */, uint64_t /* grpc_status */,
+                                 string_view /* details */) {
+    unimplemented();
+  }
+
+  // Shared Data
+
+  /**
+   * Get proxy-wide key-value data shared between VMs.
+   * @param key is a proxy-wide key mapping to the shared data value.
+   * @param cas is a number which will be incremented when a data value has been changed.
+   * @param data is a location to store the returned value.
+   */
+  virtual WasmResult getSharedData(string_view /* key */,
+                                   std::pair<std::string, uint32_t /* cas */> * /* data */);
+
+  /**
+   * Set a key-value data shared between VMs.
+   * @param key is a proxy-wide key mapping to the shared data value.
+   * @param cas is a compare-and-swap value. If it is zero it is ignored, otherwise it must match
+   * the cas associated with the value.
+   * @param data is a location to store the returned value.
+   */
+  virtual WasmResult setSharedData(string_view /* key */, string_view /* value */,
+                                   uint32_t /* cas */);
+
+  // Shared Queue
+
+  /**
+   * Register a proxy-wide queue.
+   * @param queue_name is a name for the queue. The queue_name is combined with the vm_id (if any)
+   * to make a unique identifier for the queue.
+   * @param token_ptr a location to store a token corresponding to the queue.
+   */
+  virtual WasmResult registerSharedQueue(string_view /* queue_name */, uint32_t *token_ptr);
+
+  /**
+   * Get the token for a queue.
+   * @param vm_id is the vm_id of the Plugin of the Root Context which registered the queue.
+   * @param queue_name is a name for the queue. The queue_name is combined with the vm_id (if any)
+   * to make a unique identifier for the queue.
+   * @param token_ptr a location to store a token corresponding to the queue.
+   */
+  virtual WasmResult resolveSharedQueue(string_view /* vm_id */, string_view /* queue_name */,
+                                        uint32_t * /* token_ptr */);
+
+  /**
+   * Dequeue a message from a shared queue.
+   * @param token is a token returned by registerSharedQueue();
+   * @param data_ptr is a location to store the data dequeued.
+   */
+  virtual WasmResult dequeueSharedQueue(uint32_t /* token */, std::string * /* data_ptr */);
+
+  /**
+   * Enqueue a message on a shared queue.
+   * @param token is a token returned by resolveSharedQueue();
+   * @param data is the data to be queued.
+   */
+  virtual WasmResult enqueueSharedQueue(uint32_t /* token */, string_view /* value */);
+
+  // Header/Trailer/Metadata Maps
+
+  /**
+   * Add a key-value pair to a header map.
+   * @param type of the header map.
+   * @param key is the key (header).
+   * @param value is the value (header value).
+   */
+  virtual void addHeaderMapValue(WasmHeaderMapType /* type */, string_view /* key */,
```

Comment:
For all of these functions how does one allocate a map and/or decide what map they are interacting with? If this is inferred from context, what if the map in question does not exist yet? Do you need error codes for all these functions?

Reply:
The map that your are interacting with is given by WasmHeaderMapType, e.g. WasmHeaderMapType::RequestHeaders.  We do not need error codes here for Envoy because all of the error cases have already been determined by the generic code, but I will add one in case we want to (for example) report too-many-headers.

Reply:
Done.

Reply:
So basically all of these header functions infer the header map in question in the context in question, right? I think my point still applies: what if a user tries to access a header map that doesn't exist yet? For example trying to access trailers in request headers context. Don't you need an error in that case?